api/errors/errors.go

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package errors
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"go.uber.org/yarpc/api/errors/codes"
+)
+
+// TODO: What to do with Unknown? Should we use use this for Application?
+// TODO: What to do with Internal? Should we have a function to create an error of this code?
+// TODO: should we have specific fields for each error code instead of keyValues ...string?
+
+// Code will extract the Code from a yarpc error.
+//
+// This will return None if the given error is nil or not a yarpc error.
+// This function is the defined way to test if an error is a yarpc error.
+func Code(err error) codes.Code {
+	if err == nil {
+		return codes.None
+	}
+	yarpcError, ok := err.(*yarpcError)
+	if !ok {
+		return codes.None
+	}
+	return yarpcError.Code
+}
+
+// Name will extract the user-defined error name from a yarpc error.
+//
+// This will return empty is the given error is nil, not a yarpc error, or the
+// Code is not Application.
+func Name(err error) string {
+	if err == nil {
+		return ""
+	}
+	yarpcError, ok := err.(*yarpcError)
+	if !ok {
+		return ""
+	}
+	if yarpcError.Code != codes.Application {
+		return ""
+	}
+	return yarpcError.Name
+}
+
+// Details will extract a copy the error details from a yarpc error.
+//
+// This will return empty if the given error is nil or not a yarpc error.
+func Details(err error) map[string]string {
+	if err == nil {
+		return nil
+	}
+	yarpcError, ok := err.(*yarpcError)
```

Comment:
I do this in a lot of places/it's in a lot of places in the codebase, a lot of repos violate this (which I think is good, I really dislike this directive from that wiki page).